### ISSUE TITLE
Memory leak in coroutines 1.4.x

### ISSUE THREAD

**Issue by sunpeaksfivemile** [Kotlin/kotlinx.coroutines#2564]
I've encountered a memory leak which I believe is caused by a race condition in the coroutine library. 

I've simplified the leak down to the following repro code:

```
private suspend fun <T : Any> ReceiveChannel<T>.receiveBatch(size: Int): List<T> {
    return List(size) { receive() }
}

fun main() = runBlocking {
    class Thing(val counter: Long /* for debugging on the heap dump */)

    val channel1 = produce(capacity = 64) { // from the main thread
        (0 until Long.MAX_VALUE).forEach {
            send(Thing(it))
        }
    }

    launch(newSingleThreadContext("Channel2")) { // to a single thread (This thread is important! no leak if everything is single threaded)
        flow {
            while (true) {
                emit(channel1.receiveBatch(64))
            }
        }.collect { }
    }

    Unit
}
```

This sample code leaks `Thing`s very rapidly

![Yourkit screenshot](https://i.imgur.com/Fgukd1a.png)

I haven't dug too deep, but one interesting thing is if I inline the code for receive batch, there is no leak. Example:

```
fun main() = runBlocking {
    class Thing(val counter: Long /* for debugging on the heap dump */)

    val channel1 = produce(capacity = 64) { // from the main thread
        (0 until Long.MAX_VALUE).forEach {
            send(Thing(it))
        }
    }

    launch(newSingleThreadContext("Channel2")) { // to a single thread (This thread is important! no leak if everything is single threaded)
        flow {
            while (true) {
                emit(List(64) { channel1.receive() })
            }
        }.collect { }
    }

    Unit
}
```

Tested with kotlin 1.4.30 and coroutines-core 1.4.3

**Comment by IRus** [comment]
Able to reproduce

**Comment by qwwdfsad** [comment]
Great job with a simple reproducer! Investigating this

**Comment by shmuelr** [comment]
Hey, it seems this is marked as fixed in #2581 along with a corresponding test https://github.com/Kotlin/kotlinx.coroutines/pull/2581/files#diff-ed434c42575944931c44e85218440a92177bd346ad6791cf22c45820c75413e7

Is this still an issue?
Thanks!

**Comment by qwwdfsad** [comment]
It will be fixed in an upcoming release. Version 1.4.3 still has this issue
